scala-api/src/test/scala/com/cloudera/livy/scalaapi/ScalaJobHandleTest.scala

```diff
@@ -0,0 +1,195 @@
+  }
+
+  test("onComplete Success") {
+    val jobHandleStub = new AbstractJobHandleStub[String] {
+      override def addListener(l: Listener[String]): Unit = l.onJobSucceeded(this, "hello")
+    }
+    val lock = new CountDownLatch(1)
+    var testFailure: Option[String] = None
+    val testScalaHandle = new ScalaJobHandle(jobHandleStub)
+    testScalaHandle onComplete {
+      case Success(t) => {
+        if (!t.equals("hello")) {
+          testFailure = Some("onComplete has not returned the expected message")
+        }
+        lock.countDown()
+      }
+      case Failure(e) => {
+        testFailure = Some("onComplete should not have triggered Failure callback")
+        lock.countDown()
+      }
+    }
+    if (lock.getCount == 1) lock.await()
+    testFailure.foreach(fail(_))
+  }
+
+  test("onComplete Failure") {
+    val jobHandleStub = new AbstractJobHandleStub[String] {
+      override def addListener(l: Listener[String]): Unit =
+        l.onJobFailed(this, new CustomTestFailureException)
+
+      override def get(): String = throw new CustomTestFailureException()
+    }
+    val lock = new CountDownLatch(1)
+    var testFailure: Option[String] = None
+    val testScalaHandle = new ScalaJobHandle(jobHandleStub)
+    testScalaHandle onComplete {
+      case Success(t) => {
+        testFailure = Some("Test should have thrown CustomFailureException")
+        lock.countDown()
+      }
+      case Failure(e) => {
+        if (!e.isInstanceOf[CustomTestFailureException]) {
+          testFailure = Some("Test did not throw expected exception - CustomFailureException")
+        }
+        lock.countDown()
+      }
+    }
+    if (lock.getCount == 1) lock.await()
+    testFailure.foreach(fail(_))
+  }
+
+  test("onJobCancelled") {
+    val jobHandleStub = new AbstractJobHandleStub[String] {
+      override def addListener(l: Listener[String]): Unit = l.onJobCancelled(this)
+      override def cancel(mayInterruptIfRunning: Boolean): Boolean = true
+    }
+    var testFailure: Option[String] = None
+    val lock = new CountDownLatch(1)
+    val testScalaHandle = new ScalaJobHandle(jobHandleStub)
+    testScalaHandle onJobCancelled  {
+      case true => {
+        lock.countDown()
+      }
+      case false => {
+        testFailure = Some("False callback should not have been triggered")
+        lock.countDown()
+      }
+    }
+    if (lock.getCount == 1) lock.await()
+    testFailure.foreach(fail(_))
+  }
+
+  test("onJobQueued") {
+    val jobHandleStub = new AbstractJobHandleStub[String] {
+      override def addListener(l: Listener[String]): Unit = l.onJobQueued(this)
+    }
+    var hasTestPassed = false
+    val lock = new CountDownLatch(1)
+    val testScalaHandle = new ScalaJobHandle(jobHandleStub)
+    testScalaHandle onJobQueued {
+      hasTestPassed = true
+      lock.countDown()
+    }
+    if (lock.getCount == 1) lock.await()
+    if (!hasTestPassed) fail("Callback not triggered")
```

Comment:
This check is not necessary; if the callback is not called, the previous statement will block forever. Which is why you should use a timeout in the call to `await` (and assert that it returns true).